staging/src/k8s.io/apimachinery/pkg/watch/streamwatcher.go

```diff
@@ -116,17 +123,37 @@ func (sw *StreamWatcher) receive() {
 				if net.IsProbableEOF(err) || net.IsTimeout(err) {
 					klog.V(5).Infof("Unable to decode an event from the watch stream: %v", err)
 				} else {
-					sw.result <- Event{
+					select {
+					// TODO: figure out how to test that sending the error after an externally stopped watcher
+					//       cannot block anymore because of the introduced done channel.
+					// The function receive checks under a lock whether the watch has been stopped,
+					// before an error from the watch stream is reported to the result channel.
+					// The problem here is, that in between the watcher might be stopped by
+					// calling the Stop method. In the actual code this is done by the
+					// cache.Reflector using the streamwatcher by a defer (method watchHandler)
+					// which is executed after the caller already stopped reading from the result channel.
+					// As a result the stopping flag might be set after the check
+					// and trying to send the error event blocks this send operation forever,
+					// because there will never be a receiver again.
+					// This results in dead go routines trying to send on the result channel, forever.
```

Comment:
I reread this super carefully and given that line 114 wasn't sufficient to solve the problem, are we sure that this is? In either case, there is the possibility that an event is waiting in the result channel after the Stop function is called, no?

Reply:
The channel size is one, therefore it is a synchronous exchange. Because of the common select the channel send will either be skipped if the done channel is closed in between, or aborted if the done channel is closed just after the send operation has already been started. If  the send is aborted there is no handshake and the channel is finally garbage collected.

But I think I've an even simpler solution that can even omit the `stopping` function. And I've found are a very simple test that runs on my side without false failures.
But because of the go scheduling there might be very few false accepts (test reports ok, although the error is still present). (<<0.1%). After changing from `Gosched` to 10 Milli sleeps, it even does not occur on may machine with count=10000.
I don't know why I hadn't this idea earlier. But may be some things need their time.

I'll add this with an additional commit on top.